numpy/core/tests/test_overrides.py

```diff
@@ -0,0 +1,243 @@
+        assert_equal(list(args), [])
+
+        types, args = get_overloaded_types_and_args(
+            [override_sub, no_override_sub])
+        assert_equal(set(types), {OverrideSub, NoOverrideSub})
+        assert_equal(list(args), [override_sub])
+
+    def test_ndarray_and_duck_array(self):
+
+        class Other(object):
+            __array_function__ = _return_self
+
+        array = np.array(1)
+        other = Other()
+
+        types, args = get_overloaded_types_and_args([other, array])
+        assert_equal(set(types), {np.ndarray, Other})
+        assert_equal(list(args), [other])
+
+        types, args = get_overloaded_types_and_args([array, other])
+        assert_equal(set(types), {np.ndarray, Other})
+        assert_equal(list(args), [other])
+
+    def test_many_duck_arrays(self):
+
+        class A(object):
+            __array_function__ = _return_self
+
+        class B(A):
+            __array_function__ = _return_self
+
+        class C(A):
+            __array_function__ = _return_self
+
+        class D(object):
+            __array_function__ = _return_self
+
+        a = A()
+        b = B()
+        c = C()
+        d = D()
+
+        assert_equal(_get_overloaded_args([1]), [])
+        assert_equal(_get_overloaded_args([a]), [a])
+        assert_equal(_get_overloaded_args([a, 1]), [a])
+        assert_equal(_get_overloaded_args([a, a, a]), [a])
+        assert_equal(_get_overloaded_args([a, d, a]), [a, d])
+        assert_equal(_get_overloaded_args([a, b]), [b, a])
+        assert_equal(_get_overloaded_args([b, a]), [b, a])
+        assert_equal(_get_overloaded_args([a, b, c]), [b, c, a])
+        assert_equal(_get_overloaded_args([a, c, b]), [c, b, a])
+
+
+class TestNDArrayArrayFunction(object):
+
+    def test_method(self):
+
+        class SubOverride(np.ndarray):
+            __array_function__ = _return_self
+
+        class NoOverrideSub(np.ndarray):
+            pass
+
+        array = np.array(1)
+
+        def func():
+            return 'original'
+
+        result = array.__array_function__(
+            func=func, types=(np.ndarray,), args=(), kwargs={})
+        assert_equal(result, 'original')
+
+        result = array.__array_function__(
+            func=func, types=(np.ndarray, SubOverride), args=(), kwargs={})
+        assert_(result is NotImplemented)
+
+        result = array.__array_function__(
+            func=func, types=(np.ndarray, NoOverrideSub), args=(), kwargs={})
+        assert_equal(result, 'original')
+
+
+# need to define this at the top level to test pickling
+@array_function_dispatch(lambda array: (array,))
+def dispatched_one_arg(array):
+    """Docstring."""
+    return 'original'
+
+
+class TestArrayFunctionDispatch(object):
+
+    def test_pickle(self):
+        roundtripped = pickle.loads(pickle.dumps(dispatched_one_arg))
+        assert_(roundtripped is dispatched_one_arg)
+
+    def test_docstring(self):
+        assert_equal(dispatched_one_arg.__doc__, 'Docstring.')
```

Comment:
This will need to be disabled when Python is running in `-OO` mode. It's currently causing a test failure on Travis.

Maybe it would be better to check `__name__` instead?